### PR TITLE
Deprecate `log_gpu_memory` off the Trainer constructor

### LINKED ISSUE #9032: Revamp Device Stats Logging
## Proposed refactoring or deprecation

Deprecate `log_gpu_memory` from the Trainer constructor

### Motivation

We are auditing the Lightning components and APIs to assess opportunities for improvements:
- https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

Carrying forward some of the discussion from here: https://github.com/PyTorchLightning/pytorch-lightning/discussions/9006

By default, logging gpu stats is disabled. Users must opt into this by setting. 

Issues today with this argument:
- Decoupling this from the trainer constructor gives us more flexibility to evolve the component which is actually responsible for gathering these stats without worrying about backwards compatibility guarantees for the trainer constructor argument as well. For example, the currently supported values for this flag are nvidia-smi specific. If `nvidia-smi` isn't available, this will fail. But PyTorch now supports getting CUDA memory stats directly: https://github.com/PyTorchLightning/pytorch-lightning/issues/8780 which we could take advantage of.

- Lack of consistency across different devices: We do not offer corresponding flags for TPUs nor IPUs. Creating one flag per device type would lead to further bloat in the trainer.

- There are multiple codepaths Lightning offers to log GPU memory. We have logging both [internal to the trainer](https://github.com/PyTorchLightning/pytorch-lightning/blob/13e64e6a80bc19308e99ff6aedb95ad68968b5b8/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L225-L229), as well as a [GPU stats monitor callback](https://github.com/PyTorchLightning/pytorch-lightning/blob/13e64e6a80bc19308e99ff6aedb95ad68968b5b8/pytorch_lightning/callbacks/gpu_stats_monitor.py#L38
). Having both is confusing and can lead to bugs like https://github.com/PyTorchLightning/pytorch-lightning/issues/9010 . 

### Pitch

Instead of offering separate flags & callbacks for different device types, why don't we consolidate this on the Accelerator interface, as this is where all the device-specific information already should be.

1. Offer a method like `get_device_stats` as an abstractmethod on the Accelerator interface. CPU, GPU, TPU, and IPU classes would implement this to return whatever stats are needed. This payload should be a black box to the trainer as long as it conforms to the expected return type. The implementations of this function could rely on the existing utilities we already have. This would ensure that all the device-specific code lives roughly in the same place. The stats returned should be per-device: the calling entities fetching these stats should determine whether or how to aggregate these stats (e.g. log stats from rank 0 only, reduce across ranks and log from rank 0, log the min/max across ranks, or log data from all ranks). 

2. In terms of what actually fetches these stats, we could either:
- offer a generic callback like `DeviceStatsMonitor` to periodically gather and log this data. This callback would consolidate the GPU and XLA ones and provide a general extension point for any new device types Lightning supports via the Accelerator interface. 
- Or, the trainer/loops themselves could periodically gather & log this information.

My preference is to go with the callback-based approach first. It remains the most extensible to customize options like logging frequency (e.g. step based or time-based) .

What do you think? @kaushikb11 @tchaton @awaelchli @carmocca @four4fish @edward-io @yifuwang 

### Additional context

https://github.com/PyTorchLightning/pytorch-lightning/pull/8174#discussion_r660668285
https://github.com/PyTorchLightning/pytorch-lightning/pull/9013#issuecomment-902680473

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #9032

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ananthsub** [comment]
Closing this in favor of the stack @daniellepintz is working on
